### PR TITLE
custom types

### PR DESCRIPTION
**Pull request by BastienDurel**
Add ability to define custom types, with user-defined renderers

### CODE REVIEW

**Comment by ValYouW** [comment]
Hey, looks cool, I hope to have some time later today to take a look,
did you check against jshint/jscs errors?
Thx.

**Comment by BastienDurel** [comment]
Le 18 mars 2017 13:17:09 CET, Yuval <notifications@github.com> a écrit :
>Hey, looks cool, I hope to have some time later today to take a look,
>did you check against jshint/jscs errors?
>Thx.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/ValYouW/jqPropertyGrid/pull/14#issuecomment-287541563

Yes, I did
-- 
Bastien

File: README.md [hunk 1]
```diff
@@ -45,12 +45,27 @@ var theMeta = {
...
+// This is the customTypes object that describes additionnal types, and their renderers (optional)
+var theCustomTypes = {
+	ref: {
+        name: 'ref', // name of custom type (required)
+        html: function(elemId, name, value, meta) { // custom renderer for type (required)
+            var onclick = '';
+            valueHTML = value + ' <i class="fa fa-external-link" onclick="selectRef(\'' + value + '\')"></i>';
+            return valueHTML;
+        },
+		valueFn: false // value-return function (optional). If unset, default will be "function() { return $('#' + elemId).val(); }", set to false to disable it
```
Comment (ValYouW) [inline README.md]: I think there are "tabs" here instead of "spaces" (I see extra indentation on the viewer)

File: README.md [hunk 2]
```diff
@@ -45,12 +45,27 @@ var theMeta = {
...
     framework: {name: 'Framework', group: 'Plugins', type: 'options', options: ['None', {text:'AngularJS', value: 'angular'}, {text:'Backbone.js', value: 'backbone'}], description: 'Whether to include any additional framework'},
     iAmReadOnly: { name: 'I am read only', type: 'label', description: 'Label types use a label tag for read-only properties', showHelp: false }
-    
+
+};
+
+// This is the customTypes object that describes additionnal types, and their renderers (optional)
+var theCustomTypes = {
+	ref: {
+        name: 'ref', // name of custom type (required)
```
Comment (ValYouW) [inline README.md]: Where is the `name` property is used?

File: jqPropertyGrid.js [hunk 3]
```diff
@@ -133,7 +134,7 @@
 	 * @param {function[]} [postCreateInitFuncs] - An array to fill with functions to run after the grid was created
 	 * @param {object.<string, function>} [getValueFuncs] - A dictionary where the key is the property name and the value is a function to retrieve the propery selected value
 	 */
-	function getPropertyRowHtml(pgId, name, value, meta, postCreateInitFuncs, getValueFuncs) {
+	function getPropertyRowHtml(pgId, name, value, meta, postCreateInitFuncs, getValueFuncs, customTypes) {
```
Comment (ValYouW) [inline jqPropertyGrid.js]: `customTypes` parameter is missing an entry in the function's jsdoc above...

Comment (BastienDurel) [inline jqPropertyGrid.js]: added it

File: README.md (hunk 2 above)
Comment (BastienDurel) [inline README.md]: nowhere. I guess it's legacy from first version, before I used property name. I've corrected that

**Comment by ValYouW** [comment]
@BastienDurel Thx!

**Comment by BastienDurel** [comment]
You're welcome

**Pull request merged**